### ISSUE TITLE
BaseVector::createConstant for custom types is not supported 

### ISSUE THREAD

**Issue by xiaodouchen** [facebookincubator/velox#4004]
### Bug description

I encountered the error `"Else clause of a SWITCH statement must have the same type as 'then' clauses. Expected JSON, but got VARCHAR."` when I was running a switch expression. If there is a non-null constant of json type after `then`, and a null constant of json type after `else`, the types of the two will be inconsistent after compiling by ExprCompiler.  The non-null constant of json type will be converted to Varchar type by BaseVector::createConstant(). And the null constant of json type will be converted to json type by BaseVector::createNullConstant. 

I have seen that Velox has fixed BaseVector::createNullConstant for custom types in [3849](https://github.com/facebookincubator/velox/pull/3849.). I think BaseVector::createConstant has the same problems. 

### System information

Velox System Info v0.0.2
Commit: 731831c72d4932c7ae8e8c8e69a17248738b72d7
CMake Version: 3.25.1
System: Darwin-21.4.0
Arch: arm64
C++ Compiler: /Library/Developer/CommandLineTools/usr/bin/c++
C++ Compiler Version: 13.1.6.13160021
C Compiler: /Library/Developer/CommandLineTools/usr/bin/cc
C Compiler Version: 13.1.6.13160021
CMake Prefix Path: /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr;/opt/homebrew;/usr/local;/usr;/;/opt/homebrew/Cellar/cmake/3.25.1;/usr/local;/usr/X11R6;/usr/pkg;/opt;/sw;/opt/local

**Comment by mbasmanova** [comment]
@xiaodouchen Thank you for reporting this issue. The version of BaseVector::createConstant that takes a 'variant' should be avoided as it may result in the wrong vector type (as you discovered). Can you track down the callsite and replace it with something else?

```
static std::shared_ptr<BaseVector> createConstant(
      variant value,
      vector_size_t size,
      velox::memory::MemoryPool* pool);
```

**Comment by mbasmanova** [comment]
I wonder if this is a bug in ExprCompiler:

```
ExprPtr compileExpression(

      if (constant->value().isNull()) {
        result = std::make_shared<ConstantExpr>(
            BaseVector::createNullConstant(constant->type(), 1, pool));
      } else {
        result = std::make_shared<ConstantExpr>(
            BaseVector::createConstant(constant->value(), 1, pool));
      }
```

**Comment by xiaodouchen** [comment]
Yes, this is the place. Can you help fix it? Thanks.

**Comment by mbasmanova** [comment]
I have a fix in #4020

Would you check if it fixes the problem you are seeing?

**Comment by xiaodouchen** [comment]
@mbasmanova Yes, I think it can fix my problem from the changes. Thank you!
